### PR TITLE
Tiered FS offload extension to offload kv connector

### PR DESCRIPTION
**Pull request by mattcary**
# Description

Add generic filesystem offloading.

This extends host memory offloading to a filesystem. Offloading is tiered: any chunk pushd from device to host memory is asynchronously copied to the filesystem. On lookup, if a chunk is not found in host memory, it is looked for in the filesystem. If found, it's copied to a host memory chunk and then loaded to the device.

The filesystem organization is inspired by https://github.com/llm-d/llm-d-kv-cache/tree/main/kv_connectors/llmd_fs_backend.

This has been tested on v6e to confirm functional correctness.  SGBench results are below, I can give the complete configuration on request.

### SGBench testing with Lustre

From [TPU KVCache Host Offload
benchmarks](https://docs.google.com/spreadsheets/d/12ul73Dn9TdTvS4E93GLxxe--wFuiB2XXpUzbiwDFwZo/edit?resourcekey=0-lb1vBRwpySO1tvw4NzAMNw&gid=1707653121#gid=1707653121)
and [vllm-experimental README](https://g3doc.corp.google.com/experimental/vllm-experimental/LMCache/benchmark/README.md?cl=head).

From `sglang/python`:

```
HF_TOKEN=$(< ~/.huggingface) \
python3 sglang/bench_serving.py \
  --host=localhost --port=8000 \
  --dataset-name=generated-shared-prefix \
  --model=Qwen/Qwen3-4B-Instruct-2507 --tokenizer=Qwen/Qwen3-4B-Instruct-2507 \
  --backend=vllm --gsp-num-groups=310 --gsp-prompts-per-group=40 \
  --gsp-system-prompt-len=1024 \
  --gsp-question-len=256 --gsp-output-len=256 \
  --request-rate=800 --max-concurrency=250 --seed 42
```

#### Results

fs offload, june 3 version: `--kv-transfer-config '{"kv_connector":"TPUOffloadConnector","kv_connector_module_path":"tpu_inference.offload.tpu_offload_connector","kv_role":"kv_both", "kv_connector_extra_config": {"fs_offload_base_path": "/offload"}}'`

```
============ Serving Benchmark Result ============
Backend:                                 vllm      
Traffic request rate:                    800.0     
Max request concurrency:                 250       
Successful requests:                     12400     
Benchmark duration (s):                  3729.74   
Total input tokens:                      16717720  
Total generated tokens:                  3174400   
Total generated tokens (retokenized):    3325440   
Request throughput (req/s):              3.32      
Input token throughput (tok/s):          4482.28   
Output token throughput (tok/s):         851.11    
Total token throughput (tok/s):          5333.38   
Concurrency:                             248.11    
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   74629.24  
Median E2E Latency (ms):                 60705.87  
---------------Time to First Token----------------
Mean TTFT (ms):                          72669.04  
Median TTFT (ms):                        58879.53  
P99 TTFT (ms):                           167567.69 
---------------Inter-Token Latency----------------
Mean ITL (ms):                           7.69      
Median ITL (ms):                         6.62      
P95 ITL (ms):                            12.20     
P99 ITL (ms):                            26.62     
Max ITL (ms):                            4709.72   
==================================================

(APIServer pid=8) INFO 06-03 21:54:29 [loggers.py:271] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 70.6%, External prefix cache hit rate: 98.1%
```

Same version, no fs offload: fs offload, june 3 version: `--kv-transfer-config '{"kv_connector":"TPUOffloadConnector","kv_connector_module_path":"tpu_inference.offload.tpu_offload_connector","kv_role":"kv_both"}'`

```
============ Serving Benchmark Result ============
Backend:                                 vllm      
Traffic request rate:                    800.0     
Max request concurrency:                 250       
Successful requests:                     12400     
Benchmark duration (s):                  3192.83   
Total input tokens:                      16717720  
Total generated tokens:                  3174400   
Total generated tokens (retokenized):    3320686   
Request throughput (req/s):              3.88      
Input token throughput (tok/s):          5236.01   
Output token throughput (tok/s):         994.23    
Total token throughput (tok/s):          6230.24   
Concurrency:                             247.52    
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   63733.09  
Median E2E Latency (ms):                 63853.43  
---------------Time to First Token----------------
Mean TTFT (ms):                          61704.79  
Median TTFT (ms):                        61836.77  
P99 TTFT (ms):                           72822.00  
---------------Inter-Token Latency----------------
Mean ITL (ms):                           7.96      
Median ITL (ms):                         6.62      
P95 ITL (ms):                            12.22     
P99 ITL (ms):                            71.13     
Max ITL (ms):                            736.90    
==================================================
(APIServer pid=8) INFO 06-05 00:22:07 [loggers.py:271] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 70.7%, External prefix cache hit rate: 30.1%
```

It's not in the numbers above, but lustre r/w throughput is in the 100M/s
range. I don't know if that's due to small block sizes or something, or actual
bad performance.

### CODE REVIEW

**Review by mrjunwan-lang — APPROVED** [review]

File: tpu_inference/offload/offload_manager.py [hunk 1]
```diff
@@ -126,7 +127,45 @@ def release_chunk(self, chunk: CPUChunk) -> bool:
...
+
+    @abstractmethod
+    def touch(self, chunk_hashes: list[ChunkHash]) -> int:
+        """ access chunks for both save / load; and move them to the end."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def allocate_for_save(
+        self, chunk_hashes: list[ChunkHash]
+    ) -> Tuple[list[CPUChunk], list[int]] | None:
```
Comment (saikat-royc) [inline tpu_inference/offload/offload_manager.py:147]: please add a doc string for the return args.

File: tpu_inference/offload/fs_manager.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+    def lookup(self, chunk_hashes: list[ChunkHash]) -> int:
+        """_summary_
+        return the number of cache hit starting from the first chunk
+        """
+        cpu_count = self.cpu_manager.lookup(chunk_hashes)
+
+        fs_count = 0
+        for chunk_hash in chunk_hashes[cpu_count:]:
+            # This should be async --- but in testing, this seemed to be called
```
Comment (saikat-royc) [inline tpu_inference/offload/fs_manager.py:69]: Is this comment about single hash stale? the lookup should spawn from the get_num_new_matched_tokens of kvconector scheduler   which should provide a list of block hashes
https://github.com/vllm-project/tpu-inference/blob/038f7cd34eb9a653e714ade46ea8c071bde3c97c/tpu_inference/offload/tpu_offload_connector.py#L660

Comment (mattcary) [reply tpu_inference/offload/fs_manager.py:69]: That's what I thought too, but in logs I was getting queries with just a single hash. Needs to further investigation.

File: tpu_inference/offload/fs_manager.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            # The CPU manager silently ignores hashes that aren't in the cache. We'll do
+            # the same for fs hashes, even though the lookup is expensive.
+            fs_hashes = [hash for hash in fs_hashes if self.lookup([hash])]
+
+        if fs_hashes:
+            allocated = self.cpu_manager.allocate_for_save(fs_hashes)
+            if allocated is None:
+                raise Exception("yikes, couldn't prepare load")
+            fs_chunks, chunk_indices = allocated
+            self.cpu_manager.complete_save(fs_hashes)
```
Comment (saikat-royc) [inline tpu_inference/offload/fs_manager.py:121]: by calling complete_save here we are proactively marking the fschunks ready to load?

Comment (mattcary) [reply tpu_inference/offload/fs_manager.py:121]: Exactly.

The cpu backend doesn't really care where the chunks are coming from, it's tracking a set of hash blocks in memory. They can be populated either from the accelerator, where the prepare/complete flow is called the usual way, or populated from the filesystem, where we use the save flow from the opposite direction.

File: tpu_inference/offload/fs_manager.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+            self.cpu_manager.prepare_load(fs_hashes)
```
Comment (saikat-royc) [inline tpu_inference/offload/fs_manager.py:122]: by calling complete_save() in L121, we are already calling [chunk.touch()](https://github.com/vllm-project/tpu-inference/blob/038f7cd34eb9a653e714ade46ea8c071bde3c97c/tpu_inference/offload/offload_manager.py#L229) [which increases ref count], then in L122, we are calling prepare_load() which again invokes touch again [here](https://github.com/vllm-project/tpu-inference/blob/038f7cd34eb9a653e714ade46ea8c071bde3c97c/tpu_inference/offload/offload_manager.py#L219). Is this intended?

File: tpu_inference/offload/fs_manager.py (hunk 3 above)
Comment (saikat-royc) [reply tpu_inference/offload/fs_manager.py:121]: ack. we are assuming fs reads always succeed, else we will have junk blocks in cpu backend (if the reads fs fail)?

Maybe we can put a todo here

File: tpu_inference/offload/fs_manager.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+            assert prepared
+            assert len(prepared) == len(cpu_hashes)
+            for hash, chunk in zip(cpu_hashes, prepared):
+                chunk_map[hash] = chunk
+
+        # todo: what if pulling in fs_chunks evicts cpu chunks?
+        if fs_hashes:
+            # The CPU manager silently ignores hashes that aren't in the cache. We'll do
+            # the same for fs hashes, even though the lookup is expensive.
+            fs_hashes = [hash for hash in fs_hashes if self.lookup([hash])]
```
Comment (saikat-royc) [inline tpu_inference/offload/fs_manager.py:114]: we are essentially doing 2 fs lookups. 
1. get_num_new_matched_tokens() -> manager.lookup() [which invokes fs lookups]
2. update_state_after_alloc()-> manager.prepare_load() [which again invokes fs lookups]

File: tpu_inference/offload/fs_manager.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        return res
+
+    def _start_save_to_fs(self, chunk_hash: ChunkHash, chunk: Any):
+        """
+        Async save to the backend
+        """
+        # Set the chunk size to be like the input shape, but reduced to 0.5-1 Gb per
+        # chunk, if necessary.
+        chunk_bytes = math.prod(chunk.shape)
+        factor = 750 * 2**20 / chunk_bytes
```
Comment (saikat-royc) [inline tpu_inference/offload/fs_manager.py:224]: maybe create a variable for the 750? we can also make it configurable by reading from env variable during the tieried backend init

File: tpu_inference/offload/fs_manager.py (hunk 2 above)
Comment (saikat-royc) [reply tpu_inference/offload/fs_manager.py:69]: ack

File: tpu_inference/offload/utils.py [hunk 7]
```diff
@@ -374,3 +374,62 @@ def _update_layer(cache_layer, slices):
         return jax.lax.scatter(cache_layer, indices_arr, slices, dim_nums)
 
     return jax.tree.map(_update_layer, kv_caches, layer_slices_list)
+
+
+class FileMapper:
```
Comment (saikat-royc) [inline tpu_inference/offload/utils.py:379]: can we create a separate file_mapper.py

File: tpu_inference/offload/utils.py [hunk 8]
```diff
@@ -374,3 +374,62 @@ def _update_layer(cache_layer, slices):
...
+    """
+    FileMapper maps KV blocks (given by their hash) to file names.
+
+    Inspired by llmd_fs_backend/file_mapper.py
+    """
+
+    def __init__(self, root_dir: str, vllm_config: VllmConfig):
```
Comment (saikat-royc) [inline tpu_inference/offload/utils.py:386]: note for a mulit host serving, we may need additional identifier in the path heirarchy to identify the host


cc @dannawang0221 as part of your multi host serving, did we resolve the naming convention?

Comment (dannawang0221) [reply tpu_inference/offload/utils.py:386]: What naming convention? Can you give an example? Like which worker handles the file?

File: tpu_inference/offload/tpu_offload_connector.py [hunk 9]
```diff
@@ -1597,18 +1630,20 @@ def _bucketed_update_kv_caches(
...
-            Optional[tuple[list[int], list[int]]]: A tuple containing the list of
-                source TPU blocks and destination CPU chunks if the save operation
-                is valid and contains data to be gathered. Returns None if the
+            Optional[tuple[list[int], list[int], list[ChunkHash]]: A tuple
+                containing the list of source TPU blocks, destination CPU
+                chunks, and corresponding chunk hashes if the save operation is
+                valid and contains data to be gathered. Returns None if the
                 request should be skipped or contains no new tokens to gather.
-        """
+                """
```
Comment (saikat-royc) [inline tpu_inference/offload/tpu_offload_connector.py:1642]: fix indent

File: tpu_inference/offload/tpu_offload_connector.py [hunk 10]
```diff
@@ -1311,7 +1336,14 @@ def __init__(self, vllm_config: VllmConfig,
 
         # cpu cache
         self.num_cpu_chunks = envs.TPU_OFFLOAD_NUM_CPU_CHUNKS
-        self.cpu_backend = LocalCPUBackend(num_cpu_chunks=self.num_cpu_chunks)
+        cpu_backend = LocalCPUBackend(num_cpu_chunks=self.num_cpu_chunks)
+        fs_base_path = vllm_config.kv_transfer_config.get_from_extra_config(
+            fs_manager.BASE_PATH_CONFIG, None)
+        if fs_base_path:
+            mapper = FileMapper(fs_base_path, vllm_config)
+            self.cpu_backend = fs_manager.TieredBackend(mapper, cpu_backend)
```
Comment (saikat-royc) [inline tpu_inference/offload/tpu_offload_connector.py:1344]: unfortunate variable name. but i guess we live with it, to reduce code changes :)

File: tests/offload/tpu_offload_connector_worker_test.py [hunk 11]
```diff
@@ -838,3 +859,271 @@ def test_host_memory_kind_default(self):
...
+
+        logger.info("Starting verification phase.")
+
+        assert total_num_blocks_to_save > 0
+        start = time.time()
+        while not self.cache_files() or len(
+                self.cache_files()) < total_num_blocks_to_save:
+            time.sleep(0.01)
+            assert time.time(
+            ) - start < 30, f"timeout waiting for cache files. Seeing {self.cache_files}"
```
Comment (saikat-royc) [inline tests/offload/tpu_offload_connector_worker_test.py:963]: self.cache_files()

File: tpu_inference/offload/utils.py (hunk 8 above)
Comment (saikat-royc) [reply tpu_inference/offload/utils.py:386]: cc @juncgu for mutihost seving naming convention for persistent kvchunks, which can outlive the serving workers themselves

Comment (mattcary) [reply tpu_inference/offload/utils.py:386]: Is the idea that we think of multihost serving as sharding the kv cache (one shard per host), then we need a shard id so that it loads back in correctly?

I wonder how that host id is plumbed through -- both on writing and on load.

File: tpu_inference/offload/fs_manager.py (hunk 3 above)
Comment (mattcary) [reply tpu_inference/offload/fs_manager.py:121]: If there's a load failure, then I believe _load_from_fs in the TieredBackend will raise an exception.

I'm not really sure if that's the right thing to do though...

File: tpu_inference/offload/fs_manager.py (hunk 4 above)
Comment (mattcary) [reply tpu_inference/offload/fs_manager.py:122]: Yes, that seems to be the intended flow? Look at offload_manager.py c. line 265. complete_save sets chunk.is_ready_to_load; prepare_load asserts that is_ready_to_load is true. So the intention seems to be that complete_save is called before prepare_load for a chunk (which is what would happen for cpu offload).

The fact that it's touched twice I don't quite understand --- or if I did understand I've forgotten --- but it seems to be expected?

File: tpu_inference/offload/fs_manager.py (hunk 5 above)
Comment (mattcary) [reply tpu_inference/offload/fs_manager.py:114]: Ok, I guess we should cache what's been looked up? Do we know that prepare_load is only called after get_num_new_matched_tokens?

File: tpu_inference/offload/utils.py (hunk 8 above)
Comment (saikat-royc) [reply tpu_inference/offload/utils.py:386]: As discussed offline, @mattcary  we can put a TODO here that the file mapper needs to handle TP and PP config in future iteration. The current naming conversion only supports a single host serving mode.

**Comment by saikat-royc** [comment]
lgtm from my side with minor comments. Request @juncgu  to take a look.
